lib/run.js

```diff
@@ -2,17 +2,46 @@
 
 const EE = require('events').EventEmitter
 const URL = require('url')
-const hdr = require('hdr-histogram-js')
 const timestring = require('timestring')
+const reInterval = require('reinterval')
 const Client = require('./httpClient')
 const DefaultOptions = require('./defaultOptions')
 const multipart = require('./multipart')
-const histUtil = require('hdr-histogram-percentiles-obj')
-const reInterval = require('reinterval')
+const { isMainThread } = require('./worker_threads')
 const { ofURL, checkURL } = require('./url')
 const { parseHAR } = require('./parseHAR')
-const histAsObj = histUtil.histAsObj
-const addPercentiles = histUtil.addPercentiles
+const initWorkers = require('./manager')
+const aggregateResult = require('./aggregateResult')
+const printResult = require('./printResult')
+const { getHistograms, encodeHist } = require('./histUtil')
+
+function handleFinish (result, opts, cb) {
+  if (cb) cb(null, result)
+  if (opts.json) {
+    console.log(JSON.stringify(result))
+  }
+
+  // the code below this `if` just renders the results table...
+  // if the user doesn't want to render the table, we can just return early
+  if (!isMainThread || opts.renderResultsTable === false) return
+
+  printResult(result, opts)
+}
+
+function validateOpts (opts, cb) {
+  if (opts.workers) {
+    if (opts.setupClient) {
+      cb(new Error('Can not use setupClient option while using workers'))
+      return
+    }
+
+    if (opts.requests && opts.requests.some(r => r.setupRequest || r.onResponse)) {
+      cb(new Error('Can not use setupRequest/onResponse options while using workers'))
+    }
```

Comment:
Can you give a bit more details about this limitation?

If I understand well, each worker is getting the same options (divided by the number of desired workers, when relevant).
If I ask 100 connections, with 5 workers, each will open 20 connections.

Why can't each connection invoke its own setupClient client, as it's doing when there is a single worker?

Reply:
Data sent to workers uses structured clone algorithm, which allows only limited data types. See [here](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm).

> Function objects cannot be duplicated by the structured clone algorithm; attempting to throws a DATA_CLONE_ERR exception.

And since setupRequest/onResponse are functions, we might wanna come up with some clever idea to achieve the same.
Hence, we decided to postpone that work until this PR is merged.